File: Go/bucketsort.go
package main

import "fmt"

func main() {
	fmt.Println("Enter the number of elements")
	var n int
	fmt.Scan(&n)
	arr := make([]int, n)
	fmt.Printf("Enter %v elements\n", n)
	for i := 0; i < n; i++ {
		fmt.Scan(&arr[i])
	}
	ans := BucketSort(arr)
	fmt.Println("sorted array:")
	for i := 0; i < n; i++ {
		fmt.Printf("%v ", ans[i])
	}
	fmt.Println()
}

func BucketSort(arr []int) []int {
	n := len(arr)
	max := maxval(arr)
	bucket := make([]int, max + 1)
	ans := make([]int, n)
	for i := 0; i < n; i++ {
		bucket[arr[i]]++
	}
	idx := 0
	for i := 0; i < len(bucket); i++ {
		for j := 0; j < bucket[i]; j++ {
			ans[idx] = i
			idx++
		}
	}
	return ans
}

func maxval(arr []int) int {
	max := arr[0]
	n := len(arr)
	for i := 1; i < n; i++ {
		if max < arr[i] {
			max = arr[i]
		}
	}
	return max
}

File: Php/BucketSort.php
<?php
function bucketSort(&$data)
{
    $min = $data[0];
    $max = $data[0];
    $dataLength = count($data);
    for ($i = 1; $i < $dataLength; $i++) {
        if ($data[$i] > $max) {
            $max = $data[$i];
        }

        if ($data[$i] < $min) {
            $min = $data[$i];
        }

    }
    $bucket = array();
    $bucketLength = $max - $min + 1;
    for ($i = 0; $i < $bucketLength; $i++) {
        $bucket[$i] = array();
    }
    for ($i = 0; $i < $dataLength; $i++) {
        array_push($bucket[$data[$i] - $min], $data[$i]);
    }
    $k = 0;
    for ($i = 0; $i < $bucketLength; $i++) {
        $bucketCount = count($bucket[$i]);
        if ($bucketCount > 0) {
            for ($j = 0; $j < $bucketCount; $j++) {
                $data[$k] = $bucket[$i][$j];
                $k++;
            }
        }
    }
}

$arr = array(23, 45, 1, 67, 8, 4);

bucketSort($arr);
print_r($arr);
